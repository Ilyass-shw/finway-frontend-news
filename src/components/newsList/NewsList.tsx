import React from "react";
import * as Random from "expo-random";
import { ActivityIndicator, FlatList, TouchableOpacity, View } from "react-native";
import { styles } from "./NewsListStyles";
import { useNavigation } from "@react-navigation/native";
import { NativeStackNavigationProp } from "@react-navigation/native-stack";
import { RouteParams } from "../../navigation/RootNavigator";
import ArticleItem from "../articleItem/ArticleItem";

type Source = {
  id: string;
  name: string;
};

export type article = {
  source: Source;
  author: any;
  title: string;
  description: string;
  content: string;
  url: string;
  urlToImage: string;
  publishedAt: string;
};
interface NewsListProps {
  data: article[];
  setNewsDays: React.Dispatch<React.SetStateAction<number>>;
  isLoading: boolean;
}

const NewsList: React.FC<NewsListProps> = ({
  data,
  setNewsDays,
  isLoading,
}) => {
  const navigation = useNavigation<NativeStackNavigationProp<RouteParams>>();

  const loadMoreItem = () => {
    setNewsDays((prev) => prev + 1);
  };
  const renderLoader = () => {
    return isLoading ? (
      <View style={styles.loaderStyle}>
        <ActivityIndicator size="large" color="#aaa" />
      </View>
    ) : null;
  };

  const renderListItem = ({ item }: { item: article }) => {
    const onProfilePress = () => {
      navigation.navigate("Article", { article: item });
    };
    return (
      <TouchableOpacity onPress={onProfilePress}>
        <ArticleItem article={item} />
      </TouchableOpacity>
    );
  };
  return (
    <FlatList
      style={styles.listStyle}
      data={data}
      keyExtractor={() => Random.getRandomBytes(5).toString()}
      renderItem={renderListItem}
      onEndReached={loadMoreItem}
      onEndReachedThreshold={0}
      ListFooterComponent={renderLoader}
    />
  );
};

export default NewsList;
