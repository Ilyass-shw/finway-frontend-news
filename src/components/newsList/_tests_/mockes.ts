const article1 = {
  source: { id: "eg", name: "Eng" },
  author: "Igo",
  title: "The",
  description: "Wit",
  url: "http",
  urlToImage: "https:/",
  publishedAt: "2022-01-11T23:40:24Z",
  content: "With Ma]",
};
const article2 = {
  source: { id: "eg1", name: "Eng1" },
  author: "Igo1",
  title: "The1",
  description: "Wit1",
  url: "http1",
  urlToImage: "https:/1",
  publishedAt: "2022-01-11T23:40:24Z1",
  content: "With Ma]1",
};
const article3 = {
  source: { id: "eg1", name: "Eng1" },
  author: "Igo1",
  title: "The1",
  description: "Wit1",
  url: "http1",
  urlToImage: "https:/1",
  publishedAt: "2022-01-11T23:40:24Z1",
  content: "With Ma]1",
};

export const data = [article1, article2, article3];
