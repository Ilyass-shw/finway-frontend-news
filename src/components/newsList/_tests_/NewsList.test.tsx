import { render, fireEvent } from "@testing-library/react-native";
import "@testing-library/jest-native/extend-expect";
import NewsList from "../NewsList";
import { data } from "./mockes";

jest.mock("@react-navigation/native", () => {
  const actualNav = jest.requireActual("@react-navigation/native");
  return {
    ...actualNav,
    useNavigation: () => ({
      navigate: jest.fn(),
      dispatch: jest.fn(),
    }),
  };
});

test("NewsList should render data properly", () => {
  const setNewsDays = jest.fn() as () => void;
  const { toJSON } = render(
    <NewsList data={data} setNewsDays={setNewsDays} isLoading={false} />
  );
  expect(toJSON()).toMatchSnapshot();
});

test("NewsList should display loading Icon properly", () => {
  const setNewsDays = jest.fn() as () => void;
  const { queryByTestId, rerender, debug } = render(
    <NewsList data={data} setNewsDays={setNewsDays} isLoading={false} />
  );
  expect(queryByTestId("loadingIcon")).toBe(null);
  rerender(<NewsList data={[]} setNewsDays={setNewsDays} isLoading={true} />);
  expect(queryByTestId("loadingIcon")).toBeDefined();
});
