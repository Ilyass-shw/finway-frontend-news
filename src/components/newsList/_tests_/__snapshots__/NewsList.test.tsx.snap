// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewsList should render data properly 1`] = `
<RCTScrollView
  ListFooterComponent={[Function]}
  data={
    Array [
      Object {
        "author": "Igo",
        "content": "With Ma]",
        "description": "Wit",
        "publishedAt": "2022-01-11T23:40:24Z",
        "source": Object {
          "id": "eg",
          "name": "Eng",
        },
        "title": "The",
        "url": "http",
        "urlToImage": "https:/",
      },
      Object {
        "author": "Igo1",
        "content": "With Ma]1",
        "description": "Wit1",
        "publishedAt": "2022-01-11T23:40:24Z1",
        "source": Object {
          "id": "eg1",
          "name": "Eng1",
        },
        "title": "The1",
        "url": "http1",
        "urlToImage": "https:/1",
      },
      Object {
        "author": "Igo1",
        "content": "With Ma]1",
        "description": "Wit1",
        "publishedAt": "2022-01-11T23:40:24Z1",
        "source": Object {
          "id": "eg1",
          "name": "Eng1",
        },
        "title": "The1",
        "url": "http1",
        "urlToImage": "https:/1",
      },
    ]
  }
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  onContentSizeChange={[Function]}
  onEndReached={[Function]}
  onEndReachedThreshold={0}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  style={
    Object {
      "marginHorizontal": 7,
      "overflow": "hidden",
      "width": "108%",
    }
  }
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
        testID="article wrapper"
      >
        <View
          style={
            Object {
              "alignSelf": "center",
              "backgroundColor": "white",
              "flexDirection": "row",
              "height": 160,
              "marginVertical": 4,
              "width": "90%",
            }
          }
        >
          <Image
            source={
              Object {
                "uri": "https:/",
              }
            }
            style={
              Object {
                "marginHorizontal": 9,
                "marginVertical": 12,
                "width": "50%",
              }
            }
          />
          <View
            style={
              Object {
                "width": "80%",
              }
            }
          >
            <View
              style={
                Object {
                  "height": "60%",
                  "marginTop": 14,
                  "width": "53%",
                }
              }
            >
              <Text
                ellipsizeMode="tail"
                numberOfLines={3}
                style={
                  Object {
                    "fontSize": 17,
                    "fontWeight": "700",
                  }
                }
              >
                The
              </Text>
            </View>
            <View>
              <Text
                style={
                  Object {
                    "fontSize": 14,
                    "marginTop": 8,
                  }
                }
              >
                Eng
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
        testID="article wrapper"
      >
        <View
          style={
            Object {
              "alignSelf": "center",
              "backgroundColor": "white",
              "flexDirection": "row",
              "height": 160,
              "marginVertical": 4,
              "width": "90%",
            }
          }
        >
          <Image
            source={
              Object {
                "uri": "https:/1",
              }
            }
            style={
              Object {
                "marginHorizontal": 9,
                "marginVertical": 12,
                "width": "50%",
              }
            }
          />
          <View
            style={
              Object {
                "width": "80%",
              }
            }
          >
            <View
              style={
                Object {
                  "height": "60%",
                  "marginTop": 14,
                  "width": "53%",
                }
              }
            >
              <Text
                ellipsizeMode="tail"
                numberOfLines={3}
                style={
                  Object {
                    "fontSize": 17,
                    "fontWeight": "700",
                  }
                }
              >
                The1
              </Text>
            </View>
            <View>
              <Text
                style={
                  Object {
                    "fontSize": 14,
                    "marginTop": 8,
                  }
                }
              >
                Eng1
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
        testID="article wrapper"
      >
        <View
          style={
            Object {
              "alignSelf": "center",
              "backgroundColor": "white",
              "flexDirection": "row",
              "height": 160,
              "marginVertical": 4,
              "width": "90%",
            }
          }
        >
          <Image
            source={
              Object {
                "uri": "https:/1",
              }
            }
            style={
              Object {
                "marginHorizontal": 9,
                "marginVertical": 12,
                "width": "50%",
              }
            }
          />
          <View
            style={
              Object {
                "width": "80%",
              }
            }
          >
            <View
              style={
                Object {
                  "height": "60%",
                  "marginTop": 14,
                  "width": "53%",
                }
              }
            >
              <Text
                ellipsizeMode="tail"
                numberOfLines={3}
                style={
                  Object {
                    "fontSize": 17,
                    "fontWeight": "700",
                  }
                }
              >
                The1
              </Text>
            </View>
            <View>
              <Text
                style={
                  Object {
                    "fontSize": 14,
                    "marginTop": 8,
                  }
                }
              >
                Eng1
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
    />
  </View>
</RCTScrollView>
`;
